File: MathEvaluator.py
def calc(expression):
    global exp 
    exp = expression
    exp = exp.replace(" ","")
    if find(0,"(") != -1:
        evaluateParenthesis()
    return evaluate()
    
def evaluate():
    nums = []
    ops = []
    opsi = []
    number = False
    for i in range(0,len(exp)):
        if exp[i] == '0' or exp[i] == '1' or exp[i] == '2'or exp[i] == '3'or exp[i] == '4'or exp[i] == '5'or exp[i] == '6'or exp[i] == '7'or exp[i] == '8'or exp[i] == '9':
            number = True
        elif exp[i] == '+' or exp[i] == '-' or exp[i] == '*' or exp[i] == '/':
            if number:
                ops.append(exp[i])
                opsi.append(i)
            number = False

    if len(ops) == 0:
        return float(exp)
    
    nums.append(float(exp[0:opsi[0]]))
    
    for i in range(1,len(opsi)):
        nums.append(float(exp[opsi[i-1]+1:opsi[i]]))
    
    nums.append(float(exp[opsi[len(opsi)-1]+1:len(exp)]))
    
    j = 0
    while j != len(ops):
        if ops[j] == '*':
            nums[j] = nums[j] * nums[j+1]
            del(nums[j+1])
            del(ops[j])
            j-=1
        elif ops[j] == '/':
            nums[j] = nums[j] / nums[j+1]
            del(nums[j+1])
            del(ops[j])
            j-=1
        j+=1

    j = 0
    while j != len(ops):
        if ops[j] == '+':
            nums[j] = nums[j] + nums[j+1]
            del(nums[j+1])
            del(ops[j])
            j-=1
        elif ops[j] == '-':
            nums[j] = nums[j] - nums[j+1]
            del(nums[j+1])
            del(ops[j])
            j-=1
        j+=1
    return nums[0]


def evaluateParenthesis():
    global exp
    pib = 0
    pie = len(exp)
    pf = False
    lfp = True
    neg = False
    
    while find(0,'(') != -1:
        if lfp:
            iii = 0
            while iii < len(exp):
                if exp[iii] == '(':
                    pib = iii
                iii+=1
        ii = pib+1
        while ii<len(exp):
            if exp[ii] == ')':
                pie = ii
                pf = True
                lfp = True
                ii = len(exp)
            elif exp[ii] == '(':
                pib = ii
                pf = False
                lfp = False
                ii = len(exp)
            ii+=1 
        if pf:
            if pib == 0:
                neg = False
            elif pib ==1:
                if exp[0] == '-':
                    neg = True
                else:
                    neg = False
            else:
                if (exp[pib-1] == '-') and ( exp[pib-2] == '+' or exp[pib-2] == '-' or exp[pib-2] == '*' or exp[pib-2] == '/' or exp[pib-2] == '('):
                    neg = True
                else:
                    neg = False
        if neg:
            exp = exp[0:pib-1]+str(evaluateP(pib,pie,neg))+exp[pib-1:len(exp)]
        else:
            exp = exp[0:pib]+str(evaluateP(pib,pie,neg))+exp[pib:len(exp)]

def evaluateP(i1, i2, neg):
    global exp
    nums = []
    ops = []
    opsi = []
    number = False
    for i in range(i1+1,i2):
        if exp[i] == '0' or exp[i] == '1' or exp[i] == '2'or exp[i] == '3'or exp[i] == '4'or exp[i] == '5'or exp[i] == '6'or exp[i] == '7'or exp[i] == '8'or exp[i] == '9':
            number = True
        elif exp[i] == '+' or exp[i] == '-' or exp[i] == '*' or exp[i] == '/':
            if number:
                ops.append(exp[i])
                opsi.append(i)
            number = False

    if len(ops) == 0:
        if neg:    
            value= float(exp[i1+1:i2])*-1
        else:
            value =  float(exp[i1+1:i2])
        if neg:
            exp = delete(i1-1,i2+1)
        else:
            exp = delete(i1,i2+1)
        return value
    
    nums.append(float(exp[i1+1:opsi[0]]))
    for i in range(1,len(opsi)):
        nums.append(float(exp[opsi[i-1]+1:opsi[i]]))
    nums.append(float(exp[opsi[len(opsi)-1]+1:i2]))
    

    
    
    if neg:
        exp = delete(i1-1,i2+1)
    else:
        exp = delete(i1,i2+1)
        
    i = 0
    while i!=len(ops):
        if ops[i] == '*':
            nums[i] = nums[i] * nums[i+1]
            del(nums[i+1])
            del(ops[i])
            i-=1
        elif ops[i] == '/':
            nums[i] = nums[i] / nums[i+1]
            del(nums[i+1])
            del(ops[i])
            i-=1                
        i+=1
    i = 0
    while i!=len(ops):
        if ops[i] == '+':
            nums[i] = nums[i] + nums[i+1]
            del(nums[i+1])
            del(ops[i])
            i-=1
        elif ops[i] == '-':
            nums[i] = nums[i] - nums[i+1]
            del(nums[i+1])
            del(ops[i])
            i-=1                
        i+=1
    if neg:
        return nums[0] *-1
    return nums[0]
    
def find(index,c):
    for i in range(index, len(exp)):
        if exp[i] == c:
            return i
    return -1
        
def delete(i1, i2):
    return exp[0:i1] +exp[i2:len(exp)]

File: MathEvaluator.java
// Andres Carranza
//7/22/18
// To evaluate a mathematical expression in the form of a String
import java.util.ArrayList;
import java.util.List;

public class MathEvaluator {
	private String exp;
	public double calculate(String expression) {
		//System.out.println(expression);
		exp = expression;
		deleteWhitespaces();
		evaluateParenthesis();
		return evaluate();
	}
	public double evaluate() {
		List<Double> nums = new ArrayList<Double>();
		List<Character> ops = new ArrayList<Character>();//operators
		List<Integer> opsI = new ArrayList<Integer>();//operator index
		boolean number =false;//determines if there has been a number before the last operator
		for(int i = 0;i<exp.length();i++) {
			if(exp.charAt(i)=='0'||exp.charAt(i)=='1'||exp.charAt(i)=='2'||exp.charAt(i)=='3'||exp.charAt(i)=='4'||exp.charAt(i)=='5'||exp.charAt(i)=='6'||exp.charAt(i)=='7'||exp.charAt(i)=='8'||exp.charAt(i)=='9') {
				number = true;

			}
			else if(exp.charAt(i)=='+'||exp.charAt(i)=='-'||exp.charAt(i)=='*'||exp.charAt(i)=='/') {
				if(number) {
					ops.add(exp.charAt(i));
					opsI.add(i);
				}
				number = false;
			}
		}
		if(ops.size()==0) {
			double value =Double.parseDouble(exp.substring(0, exp.length()));
			return value;
		}
		nums.add(Double.parseDouble(exp.substring(0,opsI.get(0))));
		for(int i = 1; i<opsI.size();i++) {
			nums.add(Double.parseDouble(exp.substring(opsI.get(i-1)+1,opsI.get(i))));
		}
			nums.add(Double.parseDouble(exp.substring(opsI.get(opsI.size()-1)+1,exp.length())));
			
		//after this point all values and operators have been filled in
		for(int i =0; i<ops.size();i++) {
			if(ops.get(i)=='*') {
				nums.set(i, nums.get(i)*nums.get(i+1));
				nums.remove(i+1);
				ops.remove(i);
				i--;
			}
			else if(ops.get(i)=='/') {
				nums.set(i, nums.get(i)/nums.get(i+1));
				nums.remove(i+1);
				ops.remove(i);
				i--;

			}
		}
		for(int i =0; i<ops.size();i++) {
			if(ops.get(i)=='+') {
				nums.set(i, nums.get(i)+nums.get(i+1));
				nums.remove(i+1);
				ops.remove(i);
				i--;
			}
			else if(ops.get(i)=='-') {
				nums.set(i, nums.get(i)-nums.get(i+1));
				nums.remove(i+1);
				ops.remove(i);
				i--;

			}
		}
		return nums.get(0);
	}
	public void deleteWhitespaces() {
		for(int i=0; i< exp.length();i++) {
			if(exp.charAt(i)==' ') {
				delete(i,i+1);
				i--;
			}
		}
	}
	public void evaluateParenthesis() {
		int pib =0;//( index
		int pie = exp.length();// ) index
		boolean pf = false;// true if a pair of parenthesis is found
		boolean lfp = true;//look for parenthesis
		boolean neg = false;
		while(find(0,'(')!=-1) {
			if(lfp) {
				for(int i =0;i< exp.length();i++) {
					if(exp.charAt(i)=='(') {
						pib =i;
						i = exp.length();
					}	
				}
			}
			for(int i = pib+1;i< exp.length(); i++) {
				if(exp.charAt(i)== ')') {
					pie =i;
					pf = true;
					i = exp.length();
					lfp =true;
				}
				else if(exp.charAt(i)=='(') {
					pib = i;
					i= exp.length();
					pf = false;
					lfp = false;
				}
			}
			if(pf) {
				if(pib==0)
					neg = false;
				else if(pib == 1)
					if(exp.charAt(0)=='-')
						neg = true;
					else
						neg = false;
				else
					if((exp.charAt(pib-1)=='-')&&(exp.charAt(pib-2)=='+'||exp.charAt(pib-2)=='-'||exp.charAt(pib-2)=='*'||exp.charAt(pib-2)=='/'||exp.charAt(pib-2)=='('))
						neg = true;
					else
						neg = false;
 				if(neg)
					exp = exp.substring(0, pib-1)+""+evaluateP(pib,pie,neg)+""+exp.substring(pib-1,exp.length());
				else
					exp = exp.substring(0, pib)+""+evaluateP(pib,pie,neg)+""+exp.substring(pib,exp.length());
			}

		}
	}
	public double evaluateP(int i1,int i2, boolean neg) {
		List<Double> nums = new ArrayList<Double>();
		List<Character> ops = new ArrayList<Character>();//operators
		List<Integer> opsI = new ArrayList<Integer>();//operator index
		boolean number =false;//determines if there has been a number before the last operator
		for(int i = i1+1;i<i2;i++) {
			if(exp.charAt(i)=='0'||exp.charAt(i)=='1'||exp.charAt(i)=='2'||exp.charAt(i)=='3'||exp.charAt(i)=='4'||exp.charAt(i)=='5'||exp.charAt(i)=='6'||exp.charAt(i)=='7'||exp.charAt(i)=='8'||exp.charAt(i)=='9') {
				number = true;

			}
			else if(exp.charAt(i)=='+'||exp.charAt(i)=='-'||exp.charAt(i)=='*'||exp.charAt(i)=='/') {
				if(number) {
					ops.add(exp.charAt(i));
					opsI.add(i);
				}
				number = false;
			}
		}
		if(ops.size()==0) {
			double value;
			if(neg)
				value=Double.parseDouble(exp.substring(i1+1,i2))*-1;
			else
				value=Double.parseDouble(exp.substring(i1+1,i2));
			if(neg)
				delete(i1-1,i2+1);
			else
				delete(i1,i2+1);
			return value;
		}
		nums.add(Double.parseDouble(exp.substring(i1+1,opsI.get(0))));
		for(int i = 1; i<opsI.size();i++) {
			nums.add(Double.parseDouble(exp.substring(opsI.get(i-1)+1,opsI.get(i))));
		}
		nums.add(Double.parseDouble(exp.substring(opsI.get(opsI.size()-1)+1,i2)));
		
		if(neg)
			for(int i =0; i< nums.size();i++) {
				nums.set(i, nums.get(i)*(-1));
			}
		//after this point all values and operators have been filled in
		if(neg)
			delete(i1-1,i2+1);
		else
			delete(i1,i2+1);
		for(int i =0; i<ops.size();i++) {
			if(ops.get(i)=='*') {
				nums.set(i, nums.get(i)*nums.get(i+1));
				nums.remove(i+1);
				ops.remove(i);
				i--;
			}
			else if(ops.get(i)=='/') {
				nums.set(i, nums.get(i)/nums.get(i+1));
				nums.remove(i+1);
				ops.remove(i);
				i--;

			}
		}
		for(int i =0; i<ops.size();i++) {
			if(ops.get(i)=='+') {
				nums.set(i, nums.get(i)+nums.get(i+1));
				nums.remove(i+1);
				ops.remove(i);
				i--;
			}
			else if(ops.get(i)=='-') {
				nums.set(i, nums.get(i)-nums.get(i+1));
				nums.remove(i+1);
				ops.remove(i);
				i--;

			}
		}
		return nums.get(0);
	}
	public int find(int index, char c) {
		for(int i= index; i< exp.length(); i++) {
			if(exp.charAt(i)==c)
				return i;
		}
		return -1;
	}
	public void delete(int i1, int i2) {
		//System.out.println("Before: "+ exp);
		String hold;
		hold = exp.substring(0,i1)+exp.substring(i2,exp.length());
		exp = hold;
	}
}
